include/odp/api/spec/schedule_types.h

```diff
@@ -159,6 +165,109 @@ typedef	struct odp_schedule_param_t {
 } odp_schedule_param_t;
 
+/**
+ * Scheduler capabilities
+ */
+typedef struct odp_schedule_capability_t {
+	/** Maximum number of ordered locks per queue */
+	uint32_t max_ordered_locks;
+
+	/** Maximum number of scheduling groups */
+	uint32_t max_groups;
+
+	/** Number of scheduling priorities */
+	uint32_t max_prios;
+
+	/** Scheduled queue capabilities */
+	struct {
+		/** Maximum number of scheduled (ODP_BLOCKING) queues of the
+		  * default size. */
+		uint32_t max_num;
+
+		/** Maximum number of events a scheduled (ODP_BLOCKING) queue
+		  * can store simultaneously. The value of zero means that
+		  * scheduled queues do not have a size limit, but a single
+		  * queue can store all available events. */
+		uint32_t max_size;
+
+		/** Lock-free (ODP_NONBLOCKING_LF) implementation capabilities.
+		  * The specification is the same as for the blocking
+		  * implementation. */
+		struct {
+			/** Maximum number of queues. Lock-free queues are not
+			  * supported when zero. */
+			uint32_t max_num;
+
+			/** Maximum queue size */
+			uint32_t max_size;
+
+		} lockfree;
+
+		/** Wait-free (ODP_NONBLOCKING_WF) implementation capabilities.
+		  * The specification is the same as for the blocking
+		  * implementation. */
+		struct {
+			/** Maximum number of queues. Wait-free queues are not
+			  * supported when zero. */
+			uint32_t max_num;
+
+			/** Maximum queue size */
+			uint32_t max_size;
+
+		} waitfree;
+
+		/** Maximum supported flows per queue.
+		 * Specifies the maximum number of flows per queue supported by
+		 * the implementation.  A value of 0 indicates flow aware mode
+		 * is not supported.
+		 */
+		uint32_t max_flows;
+
+	} queue;
+
+} odp_schedule_capability_t;
+
+/**
+ * Schedule configuration
+ */
+typedef struct odp_schedule_config_t {
+	/** Maximum number of schedule queues to be supported.
+	 * Application configures the maximum number of scheduled queues
+	 * required by the application.
+	 * @see odp_queue_capability_t
+	 */
+	uint32_t num_queues;
+
+	/** Scheduled queue config */
+	struct {
+		/** Maximum number of events required to be stored
+		 * simultaneously in schedule queue. This number must not
+		 * exceed 'capa.queue.max_size' required by the application.  A
+		 * value of 0 configures default queue size supported by the
+		 * implementation.
+		 */
+		uint32_t queue_size;
+
+		/** Number of flows per queue to be supported.  Scheduler
+		 * enables flow aware mode when flow count is configured
+		 * greater than 1 (up to 'capa.queue.max_flows').
+		 *
+		 * Flows are lightweight entities and events can be assigned to
+		 * specific flows by the application using
+		 * odp_event_flow_id_set() before enqueuing the event into the
+		 * scheduler. This value is ignored unless scheduler supports
+		 * flow aware mode.
+		 *
+		 * This number should be less than maximum flow supported by
+		 * the implementation.
+		 *
+		 * @see odp_schedule_capability_t
+		 */
+		uint32_t num_flows;
+	} queue;
```

Comment:
At least num and size need to be inside the same struct (blocking / LF / WF). Actually, flow support would be also per blocking level, since those may be different implementations.

In practice, HW scheduler may implement simultaneously all three blocking levels, but a SW scheduler only the first. So, it think it's fine that capa can tell if e.g. LF scheduler is available but extra config for it is not provided.

Reply:
All three params (num/size/num_flows) should be on the same level. Actually, I think we could simplify capa/config by replacing lockfree/waitfree levels, with just booleans. The reasoning is that scheduler is complex module, and would not include multiple different implementations of scheduled queues. So, either all queues are lockfree or non are. Also all params affect the implementation (so e.g. num_prios would be also a per lockfree thing, etc).

So, "struct queue" can be removed and all config params would be on the same (base) level.